Ao sq join

andioneto opened the pull request:
questions_scripts services and routes

JadziaHax committed:
Add scripts, questions, join table to init.js

The schema in init.js now creates tables for scripts, questions, and
scripts_to_questions.

JadziaHax committed:
Add scripts, questions, join table to init.js

The schema in init.js now creates tables for scripts, questions, and
scripts_to_questions.

JadziaHax committed:
Add scripts, questions, join table to init.js

The schema in init.js now creates tables for scripts, questions, and
scripts_to_questions.

JadziaHax committed:
Add scripts, questions, join table to init.js

The schema in init.js now creates tables for scripts, questions, and
scripts_to_questions.

JadziaHax committed:
Add scripts, questions, join table to init.js

The schema in init.js now creates tables for scripts, questions, and
scripts_to_questions.

JadziaHax committed:
Add catch to init.js

Added catches to each table creation call in init.js incase of rejected
promise at table creation time.

JadziaHax committed:
Add catch to init.js

Added catches to each table creation call in init.js incase of rejected
promise at table creation time.

JadziaHax committed:
Add catch to init.js

Added catches to each table creation call in init.js incase of rejected
promise at table creation time.

JadziaHax committed:
Add catch to init.js

Added catches to each table creation call in init.js incase of rejected
promise at table creation time.

JadziaHax committed:
Add catch to init.js

Added catches to each table creation call in init.js incase of rejected
promise at table creation time.

JadziaHax committed:
Add model files for questions and scripts.

JadziaHax committed:
Add model files for questions and scripts.

JadziaHax committed:
Add model files for questions and scripts.

JadziaHax committed:
Add model files for questions and scripts.

JadziaHax committed:
Add model files for questions and scripts.

andioneto committed:
basic db services for the questions model

next: will work on route controllers for questions requests

andioneto committed:
basic db services for the questions model

next: will work on route controllers for questions requests

andioneto committed:
basic db services for the questions model

next: will work on route controllers for questions requests

andioneto committed:
basic db services for the questions model

next: will work on route controllers for questions requests

andioneto committed:
basic db services for the questions model

next: will work on route controllers for questions requests

andioneto committed:
basic db services for the questions model

next: will work on route controllers for questions requests

andioneto committed:
basic db services for the questions model

next: will work on route controllers for questions requests

andioneto committed:
basic db services for the questions model

next: will work on route controllers for questions requests

andioneto committed:
basic db services for the questions model

next: will work on route controllers for questions requests

andioneto committed:
fixed some minor lint warnings

andioneto committed:
fixed some minor lint warnings

andioneto committed:
fixed some minor lint warnings

andioneto committed:
fixed some minor lint warnings

andioneto committed:
fixed some minor lint warnings

andioneto committed:
fixed some minor lint warnings

andioneto committed:
fixed some minor lint warnings

andioneto committed:
fixed some minor lint warnings

andioneto committed:
fixed some minor lint warnings

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

andioneto committed:
added routes, controllers and db services for questions
also updated middleware to include questions route

next: work with jadzia on join table between scripts and questions

JadziaHax committed:
Add controller, service, and test for scripts.

Adds the controller, services and test for scripts. Scripts can be
created, fetched singly by id or by all, updated by id, and destroyed by
id.

JadziaHax committed:
Add controller, service, and test for scripts.

Adds the controller, services and test for scripts. Scripts can be
created, fetched singly by id or by all, updated by id, and destroyed by
id.

JadziaHax committed:
Add controller, service, and test for scripts.

Adds the controller, services and test for scripts. Scripts can be
created, fetched singly by id or by all, updated by id, and destroyed by
id.

JadziaHax committed:
Add controller, service, and test for scripts.

Adds the controller, services and test for scripts. Scripts can be
created, fetched singly by id or by all, updated by id, and destroyed by
id.

JadziaHax committed:
Add controller, service, and test for scripts.

Adds the controller, services and test for scripts. Scripts can be
created, fetched singly by id or by all, updated by id, and destroyed by
id.

JadziaHax committed:
Add controller, service, and test for scripts.

Adds the controller, services and test for scripts. Scripts can be
created, fetched singly by id or by all, updated by id, and destroyed by
id.

andioneto committed:
basic outline of tests for the questions services

andioneto committed:
basic outline of tests for the questions services

andioneto committed:
basic outline of tests for the questions services

andioneto committed:
basic outline of tests for the questions services

andioneto committed:
basic outline of tests for the questions services

andioneto committed:
basic outline of tests for the questions services

andioneto committed:
basic outline of tests for the questions services

andioneto committed:
basic outline of tests for the questions services

andioneto committed:
added db config to the questions model file

andioneto committed:
added db config to the questions model file

andioneto committed:
added db config to the questions model file

andioneto committed:
added db config to the questions model file

andioneto committed:
added db config to the questions model file

andioneto committed:
added db config to the questions model file

andioneto committed:
added db config to the questions model file

andioneto committed:
removed unnecessary import of scripts service

andioneto committed:
removed unnecessary import of scripts service

andioneto committed:
removed unnecessary import of scripts service

andioneto committed:
removed unnecessary import of scripts service

andioneto committed:
Merge branch 'ao-quest-services-cntrls' of https://github.com/Twilio-org/phonebank into ao-quest-services-cntrls

JadziaHax committed:
Add saveNewScript to script controller. Remove unused import from model.

JadziaHax committed:
Add saveNewScript to script controller. Remove unused import from model.

andioneto committed:
debugging questions service testing + table not being created wip

andioneto committed:
debugging questions service testing + table not being created wip

andioneto committed:
debugging questions service testing + table not being created wip

andioneto committed:
debugging questions service testing + table not being created wip

andioneto committed:
debugging questions service testing + table not being created wip

andioneto committed:
debugging questions service testing + table not being created wip

andioneto committed:
debugging questions service testing + table not being created wip

andioneto committed:
debugging questions service testing + table not being created wip

andioneto committed:
Merge branch 'ao-quest-services-cntrls' of https://github.com/Twilio-org/phonebank into ao-quest-services-cntrls

andioneto committed:
basic outline for questions db service tests done

andioneto committed:
basic outline for questions db service tests done

andioneto committed:
basic outline for questions db service tests done

andioneto committed:
basic outline for questions db service tests done

andioneto committed:
updated route callback functions for questions to relfect name changes in the route controllers

andioneto committed:
updated route callback functions for questions to relfect name changes in the route controllers

andioneto committed:
updated route callback functions for questions to relfect name changes in the route controllers

andioneto committed:
updated route callback functions for questions to relfect name changes in the route controllers

andioneto committed:
Merge branch 'ao-quest-services-cntrls' of https://github.com/Twilio-org/phonebank into ao-quest-services-cntrls

JadziaHax committed:
Add Scripts Model with service, controller, router, and tests.

Adds the scripts model with the relevant service, controller, router,
and test files.

JadziaHax committed:
Add Scripts Model with service, controller, router, and tests.

Adds the scripts model with the relevant service, controller, router,
and test files.

JadziaHax committed:
Remove references to questions in script router.

JadziaHax committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into zia-model-solo

andioneto committed:
basic outline for scripts_to_questions services and controllers

andioneto committed:
basic outline for scripts_to_questions services and controllers

andioneto committed:
Merge branch 'zia-model-solo' of https://github.com/Twilio-org/phonebank into ao-sq-join

andioneto committed:
debugging join table services and controllers for questions_scripts

andioneto committed:
debugging join table services and controllers for questions_scripts

andioneto committed:
bug fixed, able to insert into the questions_scripts join table

note: changed name of join table to questions_scripts as it was necessary for functionality
      also added the name 'id' to scripts and questions primary key, necessary for query

next: testing for this service + controller

andioneto committed:
bug fixed, able to insert into the questions_scripts join table

note: changed name of join table to questions_scripts as it was necessary for functionality
      also added the name 'id' to scripts and questions primary key, necessary for query

next: testing for this service + controller

andioneto committed:
fixed CI problems

andioneto committed:
fixed CI problems

andioneto committed:
Merge branch 'ao-sq-join' of https://github.com/Twilio-org/phonebank into ao-sq-join

perfectlynormalbeast commented on server/db/services/scripts.js:
```diff
@@ -0,0 +1,58 @@
+
```
nit: 2 bogus newlines

JadziaHax replied:
will remove

JadziaHax commented on test/server/services/questions.js:
```diff
@@ -0,0 +1,142 @@
+import Model from '../../../server/db/models/questions';
+
+const knex = knexModule(testconfig);
+const knexdb = bookshelfModule(knex);
```
this object should probably be called bookshelf. I know Duy used this pattern but its an artifact of a time when there was an extemporaneous import of bookshelf.

perfectlynormalbeast commented on server/db/services/scripts.js:
```diff
@@ -0,0 +1,58 @@
+  getQuestionsByScriptId: (params, Model) => {
+    const { id } = params;
+
+    return new Model({ id }).questions().fetchAll();
```
Perhaps also order by sequence number?

JadziaHax replied:
agreed

andioneto replied:
got it!

JadziaHax committed:
Change const 'knexdb' to 'bookshelf' (name-only change).

andioneto committed:
linting fixes

perfectlynormalbeast commented on server/controllers/questions.js:
```diff
@@ -0,0 +1,58 @@
+import knexModule from 'knex';
+import bookshelfModule from 'bookshelf';
+import questionsService from '../db/services/questions';
+import { development as devconfig } from '../../knexfile';
```
This is actually a bad pattern - ties the code to a dev config. A production deployment should not require a code change. Is there a way for the main app to detect dev/prod environment and ideally pass in a preinitialized bookshelf object that is shared between all controllers? This will eliminate the need for lines 9 and 10.

andioneto replied:
I think this has been resolved. We can talk about this more indepth as a group though

perfectlynormalbeast commented on server/controllers/questions.js:
```diff
@@ -0,0 +1,58 @@
+import { development as devconfig } from '../../knexfile';
+import Question from '../db/models/questions';
+
+const { saveNewQuestion, getAllQuestions, getQuestionById } = questionsService;
```
Can we do this in line 3 itself?

andioneto replied:
My bad, fixed!

perfectlynormalbeast commented on server/controllers/scripts.js:
```diff
@@ -0,0 +1,136 @@
+import scriptsService from '../db/services/scripts';
+import { development as devconfig } from '../../knexfile';
+import Script from '../db/models/scripts';
+// import bookshelfCreateTable from '../db/init';
```
nit: remove unused line

andioneto replied:
done!

perfectlynormalbeast commented on server/controllers/scripts.js:
```diff
@@ -0,0 +1,136 @@
+import knexModule from 'knex';
```
Same comments about knex, bookshelf and config as above.

perfectlynormalbeast commented on server/controllers/scripts.js:
```diff
@@ -0,0 +1,136 @@
+    })
+    .catch((err) => {
+      console.log(err);
+      res.status(401).json({ message: 'Could not save script.' });
```
Why are we returning a 401 here? In fact, other route methods don't seem to handle errors at all? What is the default behavior for express when a route method doesn't return a value or errors out?

andioneto replied:
@JadziaHax This might be a better question for you.

@perfectlynormalbeast From what I can see, most of the error handling is in middleware.

JadziaHax replied:
The default behavior is to return a 500 error, usually the console error trace formatted as an HTML page.

JadziaHax replied:
I'm removing the 401 as that was included by mistake.

perfectlynormalbeast commented on server/config/routes/scripts.js:
```diff
@@ -0,0 +1,21 @@
+router.route('/:id').get(getScriptById);
+router.route('/:id').put(updateScriptById);
+router.route('/:id').delete(deleteScriptById);
+router.route('/addQuestionToScript/:id/:questionId/:sequenceNum').post(addQuestionToScript);
```
Would consider changing the route for this and the one below to `:id/questions` and make it accept form params on POST for adding a question and a plain GET to get questions. Makes for a cleaner routespace for our server.

andioneto replied:
`router.route('/:id/scriptQuestions/:questionId/:sequenceNum').post(addQuestionToScript);
router.route('/:id/scriptQuestions/').get(getQuestionsByScriptId);`

like that?

perfectlynormalbeast replied:
No for the create relationship route, I meant `router.route('/:id/scriptQuestions/` and make the other 2 params as JSON/form params in the POST.
The second route looks good.

perfectlynormalbeast commented on server/db/models/questions.js:
```diff
@@ -0,0 +1,20 @@
+
+import Script from './scripts';
+
+import { development as devconfig } from '../../../knexfile';
```
Same comment about knex, bookshelf and config as above.

perfectlynormalbeast commented on server/db/models/scripts.js:
```diff
@@ -0,0 +1,19 @@
+import knexModule from 'knex';
```
Same comment as knex, bookshelf and config as above.

perfectlynormalbeast commented on test/server/services/questions.js:
```diff
@@ -0,0 +1,142 @@
+const bookshelf = bookshelfModule(knex);
+const questionsModel = Model(bookshelf);
+
+// const assert = chai.assert;
```
Nit: remove commented out unused lines

andioneto replied:
✔️

perfectlynormalbeast commented on server/db/init.js:
```diff
@@ -22,7 +22,53 @@ bookshelf.knex.schema.hasTable('users').then((exist) => {
+  if (!exist) {
+    bookshelf.knex.schema.createTable('questions', (table) => {
+      table.increments('id').primary();
+      table.string('title').notNullable().index();
```
I think this should be unique. The change may break some tests.

perfectlynormalbeast replied:
Bump - let's make this unique?

perfectlynormalbeast commented on server/db/init.js:
```diff
@@ -22,7 +22,53 @@ bookshelf.knex.schema.hasTable('users').then((exist) => {
+  if (!exist) {
+    bookshelf.knex.schema.createTable('scripts', (table) => {
+      table.increments('id').primary();
+      table.string('name').notNullable().index();
```
I think this should be unique. The change may break some tests.

andioneto replied:
Yeah, we still need to speak to the team about which fields we want to require unique values, I don't think we got to that yesterday.

perfectlynormalbeast replied:
Bump - let's make this unique?

perfectlynormalbeast reviewed (changes requested):
Left a couple of questions. Including some high-level concerns about re-initializing knex+bookshelf+config for each controller.

andioneto committed:
addressed comments on pull request

need to sync with Duy to discuss models

andioneto committed:
Merge branch 'ao-sq-join' of https://github.com/Twilio-org/phonebank into ao-sq-join

JadziaHax committed:
Add Tests for scripts_questions services

This adds tests for scripts_questions services. It also modifies the
addQuestionToScript function to return the script_id of the relation,
and removes non-functional authorization filter on questions routes.

JadziaHax committed:
Merge branch 'ao-sq-join' of https://github.com/Twilio-org/phonebank into ao-sq-join

andioneto committed:
fixed query for getQuestionsByScriptId, now returning question information, ordered by sequence_number

JadziaHax committed:
Remove 401 status in script controller error handle.

andioneto committed:
Merge branch 'ao-sq-join' of https://github.com/Twilio-org/phonebank into ao-sq-join

perfectlynormalbeast commented on server/db/services/scripts.js:
```diff
@@ -42,15 +42,22 @@ export default {
     const { id } = params;
-
-    return new Model({ id }).questions().fetchAll().orderBy('sequence_number', 'DESC');
+    return new Model().query((q) => {
```
Why are we needing to construct this query manually?

perfectlynormalbeast commented on test/server/services/questionsScripts.js:
```diff
@@ -0,0 +1,207 @@
+        }, done);
+    });
+
+    it('the saved join-entry only has 6 attributes after saving', (done) => {
```
nit: 3 attributes, not 6

perfectlynormalbeast commented on test/server/services/questionsScripts.js:
```diff
@@ -0,0 +1,207 @@
+
+      scriptsService.getQuestionsByScriptId(params, ScriptsModel)
+        .then((questions) => {
+          console.log(questions);
```
This isn't a test, this is just logging the questions.

JadziaHax committed:
Fix typo in test language for questions_scripts.

JadziaHax committed:
Merge branch 'ao-sq-join' of https://github.com/Twilio-org/phonebank into ao-sq-join

andioneto committed:
modified scriptsQuestions route to accept a script id in the route params and question id + sequence number in the body of the request
also some minor linting issues

JadziaHax committed:
Modify property names in test file to make them pass again

This also adds script_id to the addQuestionToScript function in the
scripts service as has been done previously, but then later undone.

JadziaHax committed:
Merge branch 'ao-sq-join' of https://github.com/Twilio-org/phonebank into ao-sq-join

JadziaHax committed:
Fix promise chaining to use Promise.all where possible.

JadziaHax committed:
Add console.log for checking Travis config

JadziaHax committed:
Move config console output to questisons_scripts test instead of users.

JadziaHax committed:
Remove questions_scripts tests until we can refactor database config.

These tests will be brought back when we have a database configuration
that doesn't depend on a hardcoded config being passed every time we
need access to a model.

JadziaHax committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-sq-join

perfectlynormalbeast commented on server/db/init.js:
```diff
@@ -22,7 +23,53 @@ bookshelf.knex.schema.hasTable('users').then((exist) => {
+
+bookshelf.knex.schema.hasTable('questions_scripts').then((exist) => {
+  if (!exist) {
+    bookshelf.knex.schema.createTable('questions_scripts', (table) => {
```
Maybe add a unique constraint for script_id and question_id?
`table.unique(['script_id', 'question_id])`

perfectlynormalbeast reviewed (changes requested):
Some small changes

JadziaHax committed:
Add unique constraints to scripts, questions, and questions_scripts

perfectlynormalbeast commented on server/db/init.js:
```diff
@@ -46,8 +45,8 @@ bookshelf.knex.schema.hasTable('scripts').then((exist) => {
     bookshelf.knex.schema.createTable('questions', (table) => {
-      table.increments('id').primary();
-      table.string('title').notNullable().index();
+      table.increments('id').primary().unique();
```
Unnecessary unique constraint - primary key implies uniqueness.

JadziaHax committed:
Remove unintended unique constraint on the id column of questions.

perfectlynormalbeast committed:
Merge branch 'master' into ao-sq-join
